cryptofeed/feedhandler.py

```diff
@@ -227,21 +231,49 @@ def run(self, start_loop: bool = True, install_signal_handlers: bool = True):
                 loop.create_task(self._connect(conn, sub, handler))
                 self.timeout[conn.uuid] = timeout
 
-        if start_loop:
-            try:
-                loop.run_forever()
-            except SystemExit:
-                LOG.info("System Exit received - shutting down")
-            except Exception:
-                LOG.error("Unhandled exception", exc_info=True)
-            finally:
-                self.stop(loop=loop)
+        if not start_loop:
+            return
+
+        try:
+            loop.run_forever()
+        except SystemExit:
+            LOG.info('FH: System Exit received - shutting down')
+        except Exception as why:
+            LOG.exception('FH: Unhandled %r - shutting down', why)
+        finally:
+            self.stop(loop=loop)
 
     def stop(self, loop=None):
         if not loop:
             loop = asyncio.get_event_loop()
+
+        LOG.info('FH: create the tasks to properly shutdown the backends (data flush)')
+        shutdown_tasks = []
         for feed, _ in self.feeds:
-            loop.run_until_complete(feed.stop())
+            task = loop.create_task(feed.shutdown())
+            task.set_name(f'shutdown_feed_{feed.id}')
+            shutdown_tasks.append(task)
+
+        LOG.info('FH: wait %s backend tasks until termination', len(shutdown_tasks))
+        try:
+            loop.run_until_complete(asyncio.gather(*shutdown_tasks))
+        finally:
+            pass
+
+        LOG.info('FH: stop the AsyncIO loop: wait for the current batch of callbacks and then exit')
+        loop.stop()
```

Comment:
I have some questions about the code below this point. 

A common request from users of the library is to do loop mgmt themselves, so I'm not sure we should be calling stop(), close(), or shutdown_asyncgens() here for every shutdown. What we could do is optionally do these things based on if the loop is provided. A user who is managing their own loop should supply the loop themselves (or it could be made a requirement to ensure these are not called). Another option might be to check if we are stopping because of a signal or because a user called the stop method. Thoughts?

Reply:
I also thought that we could add yet another kwarg, something like "shutdown_loop" and default it to True. Thats probably easiest

Reply:
Yep sure. I ignored users that manages themself the ASyncIO event loop and want to call the `FeedHandler.stop()` only to flush the local cache data.
 
My proposition is to move the extra code of the function `stop()` into a new function `close()`.
The user can call the function `stop()` to only shutdown the backends, and may also call the function `close()` is desired.

Please se my last commit. :warning: To be tested.